### ISSUE TITLE
The ghc -M / -c approach is not sufficient for correct builds in all cases

### ISSUE THREAD

**Issue by nh2** [23Skidoo/ghc-parmake#9]
See: http://ghc.haskell.org/trac/ghc/ticket/481

While there is a [way to get around the external dependencies problem](https://github.com/23Skidoo/ghc-parmake/pull/8), the cases regarding TemplateHaskell described in that bug can probably not be implemented using `ghc -M` and `ghc -c`.

(Also, http://ghc.haskell.org/trac/ghc/ticket/7277 is not even handled properly by `ghc --make` itself.)

This is probably important for the GSoC project, because it means that any system based on `ghc -M` alone cannot produce always-correct builds.

The only workaround that I see for now is to call `ghc --make` after our parallel build, and live with the singlethreadedness that comes with that.

**Comment by ezyang** [comment]
Why not write some programs using the GHC API to help you get out the information you need? You can define a private interface between ghc-parmake and the GHC API program which is shared across versions of GHC, so you can have different versions of the GHC API program for different versions of GHC. GHC 8.0 is even coming with frontend plugins (https://ghc.haskell.org/trac/ghc/ticket/11194) which should make it even easier to write programs which accept GHC-style arguments.

Once you figure out exactly what you need, and what a good format is, you can submit the frontends to GHC for upstreaming.

**Comment by 23Skidoo** [comment]
Thanks, I'll try to find some time to experiment with frontend plugins. 

The master branch actually contains a GHC API-based reimplementation of `ghc-parmake` (which is not up to par with the 0.1 branch yet).

**Comment by ezyang** [comment]
Yeah, the GHC API can be really helpful. For example, in https://github.com/ezyang/ghc-shake I can run the preprocessor and parseHeader, get the dependencies, and then feed it directly into the build system. For this approach to scale, I need to separate the GHC API bits from the build system bits, but it works _very_ nicely.

**Comment by 23Skidoo** [comment]
> For example, in https://github.com/ezyang/ghc-shake I can run the preprocessor and parseHeader, get the dependencies, and then feed it directly into the build system.

Looks pretty cool!
